### PR TITLE
Add saving of test predictions after training 

### PR DESCRIPTION
**Pull request by am2145**
## Description
This PR adds functionality for the CLI argument `--save-test-preds` in `train.py`. The predictions on the test set after training are saved in the output directory to a CSV file `test_predictions.csv`. I have also updated the regression CLI test to check that the saved predictions CSV exists in line with the rest of the test.

## Questions
I have detailed the open questions in issue #662 and have made an implementation based on my discussion with @KnathanM there. 

1) I added a function `get_column_names` to `parsing.py` to retrieve the column names that we'd use when saving the predictions, but I'm worried there is too much duplication with `parse_csv`. Is there a better implementation approach? 

2) I added a simple CLI test with @hwpang that checks if a `predictions.csv` file is written but nothing beyond the file is existing. Do we want a test to check that the `predictions.csv` file is written with the correct column headers and values? 

## Relevant issues
#662: Essentially, we would like to have column headers for the smiles string as well as to say what target each of the predictions correspond to, but the user might not always provide this as input. We have tried to cover the various cases of what might be provided in the input csv file, but there are open questions on the implementation.    

## Checklist
- [x] linted with flake8?
- [x] (if appropriate) unit tests added?

### CODE REVIEW

File: chemprop/cli/train.py [hunk 1]
```diff
@@ -3,6 +3,7 @@
 from pathlib import Path
 import sys
 import json
+import numpy as np
```
Comment (KnathanM) [inline chemprop/cli/train.py:6]:
```suggestion
```

File: chemprop/cli/train.py [hunk 2]
```diff
@@ -733,11 +734,18 @@ def train_model(args, train_loader, val_loader, test_loader, output_dir, scaler)
...
+                model.predictor.register_buffer("scale", torch.tensor(scaler.scale_).view(1, -1))
             results = trainer.test(model, test_loader)[0]
             logger.info(f"Test results: {results}")
 
+            if args.save_preds:
+                preds = trainer.predict(model, test_loader)
+                preds_list = [pred.flatten() for pred in np.rollaxis(torch.stack(preds).numpy(),-1)]
+                columns = get_column_names(args.data_path, args.smiles_columns, args.reaction_columns, args.target_columns, args.ignore_columns, args.no_header_row)
+                df_preds = pd.DataFrame(list(zip(test_loader.dataset.smiles, *preds_list)), columns = columns)
+                df_preds.to_csv(model_output_dir / "test_predictions.csv", index=False)
```
Comment (KnathanM) [inline chemprop/cli/train.py:747]:
```suggestion
                predss = trainer.predict(model, test_loader)
                preds = torch.concat(predss, 0).numpy()
                columns = get_column_names(args.data_path, args.smiles_columns, args.reaction_columns, args.target_columns, args.ignore_columns, args.no_header_row)
                df_preds = pd.DataFrame(list(zip(test_loader.dataset.smiles, *preds.T)), columns = columns)
                df_preds.to_csv(model_output_dir / "test_predictions.csv", index=False)
```
This makes the style consistent with `predict.py`.

**Review by KnathanM — COMMENTED** [review]
I discussed this PR with Angiras. It seems unclear whether we actually need a save-preds argument because the user can request their splits and then run prediction manually on the test split. But this was in v1 so there was someone who thought it was useful. And if you have multiple targets, doing this in train allows for the prediction column titles to be related to the target column titles. 

We also considered whether save-preds could just reuse a function from `predict.py`, but found that would be difficult. They are similar in that they both use `trainer.predict` to get predictions and `df.to_csv` to save the predictions, but in `predict.py` there is no hope of automatically giving descriptive names to the prediction columns (as the input only has SMILES) which that is possible in `train.py`. Additionally in `train.py` the SMILES are taken from `test_loader.dataset.smiles` because unlike `predict.py` only a subset of the input file will get predictions.

File: chemprop/cli/train.py (hunk 2 above)
Comment (am2145) [reply chemprop/cli/train.py]: This is much cleaner and removes the need to specifically import numpy. I have committed the suggested changes.

**Review by KnathanM — APPROVED** [review]
